ddtrace/settings/_inferred_base_service.py

```diff
@@ -0,0 +1,281 @@
+                - bool: A flag indicating whether valid metadata was found.
+        """
+        prev_arg_is_flag = False
+        module_flag = False
+
+        for arg in args:
+            has_flag_prefix = arg.startswith("-")
+            is_env_variable = "=" in arg
+
+            should_skip_arg = prev_arg_is_flag or has_flag_prefix or is_env_variable
+
+            if module_flag:
+                return ServiceMetadata(arg), True
+
+            if not should_skip_arg:
+                abs_path = pathlib.Path(arg).resolve()
+                if not abs_path.exists():
+                    continue
+                stripped = abs_path
+                if not stripped.is_dir():
+                    stripped = stripped.parent
+                value, ok = self.deduce_package_name(stripped)
+                if ok:
+                    return ServiceMetadata(value), True
+                return ServiceMetadata(self.find_nearest_top_level(stripped)), True
+
+            if has_flag_prefix and arg == "-m":
+                module_flag = True
+
+            prev_arg_is_flag = has_flag_prefix
+
+        return ServiceMetadata(""), False
+
+    def deduce_package_name(self, fp: pathlib.Path) -> Tuple[str, bool]:
+        # Walks the file path until a `__init__.py` is not found.
+        # All the dir traversed are joined then with `.`
+        up = pathlib.Path(fp).parent
+        current = fp
+        traversed: List[str] = []
+
+        while current != up:
+            if not (current / INIT_PY).exists():
+                break
+            traversed.insert(0, current.name)
+            current = up
+            up = current.parent
+
+        return ".".join(traversed), len(traversed) > 0
+
+    def find_nearest_top_level(self, fp: pathlib.Path) -> str:
+        # returns the top level dir containing a .py file starting walking up from fp
+        up = fp.parent
+        current = fp
+        last = current
+
+        while current != up:
+            if not fnmatch.filter(os.listdir(current), ALL_PY_FILES):
+                break
+            last = current
+            current = up
+            up = current.parent
+
+        return last.name
+
+
+class GunicornDetector:
+    def __init__(self, ctx):
+        self.ctx = ctx
+        name = "gunicorn"
+        self.name = name
+        self.pattern = name
+
+    def expected_command_name(self) -> str:
+        return "gunicorn"
+
+    def detect(self, args: List[str]) -> Tuple[ServiceMetadata, bool]:
+        """
+        Extracts and returns the Gunicorn application name from various potential sources.
+
+        This function first checks for the application name in the environment variable
+        specified by `GUNICORN_CMD_ARGS`. If found, it extracts the name from it.
+        Next, it looks for the WSGI application name in the environment variable
+        `WSGI_APP_ENV`. If that is not present, it tries to extract the name from
```

Comment:
Should we handle the service naming on the integration level?  Web frameworks (ex: [Flask](https://flask.palletsprojects.com/en/stable/tutorial/factory/#the-application-factory). [django](https://docs.djangoproject.com/en/5.1/topics/settings/#envvar-DJANGO_SETTINGS_MODULE), [fastapi](https://fastapi.tiangolo.com/advanced/settings/#the-env-file)) have their own way of setting application names. It's not necessarily done by environment variable. Some web frameworks also have a programatic API for setting application names. This will approach will completely miss this.  

Also the ddtrace library does not provide tracing for http servers directly (ex: gunicorn, Uvicorn, Uwsgi, hypercorn, Waitress). We only support tracing for the web frameworks run on these servers. This change introduces a new layer of instrumentation that increases the complexity of our integrations.